Finance Template Documentation (#761)
ethereum-optimism/superchain-ops: src/improvements/doc/FINANCE_TEMPLATE.md

# Finance Template Usage

The Finance Template is used for approving or sending ETH or ERC20 tokens from a Safe, and allows multiple actions to be batched into a single transaction.

Each Finance Template transaction requires developers to create a new `.toml` file containing the following parameters.

See [Single](../SINGLE.md) and [Nested](../NESTED.md) for details on running tasks.

## Example

```toml

> # Each Finance Template transaction requires developers to create a new `.toml` file containing the following parameters.

Reviewer: ```suggestion
# Usage of the Finance Template transaction requires creating a new `.toml` file containing the following parameters.
# This should be done using the `just new task` command
```

# Finance Template Usage

> The Finance Template allows members of the finance team to send tokens and Ether to multiple addresses in a single transaction.

Reviewer: nit here in that anyone can use this, not just people on the finance team
```suggestion
The Finance Template is used for approving or sending ETH or ERC20 tokens from a Safe, and allows multiple actions to be batched into a single transaction.
```

> ## Usage

Reviewer: Let's just copy in a sample TOML file and annotate with comments, e.g. something like this: https://github.com/ethereum-optimism/superchain-ops/pull/739#discussion_r1989893296

I think that will be more clear since the full context of the TOML to work with is visible while reading the docs

Reviewer: Ah nice I see the example below. Yea let's just make that the primary docs and add everything else here as comments

Each Finance Template transaction requires developers to create a new `.toml` file containing the following parameters.

**Template Name** - This specifies which template to use. For the Finance Template, the template name is `FinanceTemplate`.

```templateName = "FinanceTemplate"```

> **Operation Type** - This can be of type `approve`, `increaseAllowance`, `decreaseAllowance`, or `transfer`. The operation type specifies the type of operation to perform on the token being transferred. If the token is Eth, the operation type can only be `transfer`.

Reviewer: ```suggestion
**Operation Type** - This can be of type `approve`, `increaseAllowance`, `decreaseAllowance`, or `transfer`. The operation type specifies the type of operation to perform on the token being transferred. If the token is ETH, the operation type can only be `transfer`.
```

```operationType = "approve"```

- **Operations** - The operations array specifies the amounts, tokens, and recipients for each transfer. Each operation object has the following parameters:

>     - **token** - The identifier of the token to transfer. This should be the identifier of the token address as specified in `addresses`.

Reviewer: ```suggestion
    - **token** - A string representing the token to transfer. This string corresponds to the token's key in the `addresses` section
```

>     - **amount** - The amount of the token to transfer to the recipient.

Reviewer: ```suggestion
    - **amount** - The quantity of the token to use in the operation. For transfers this is the amount transferred, for approvals is the new approval amount, etc. This is specified as a human-readable value, not using the full amount of decimals for a token. For example, `1` for ETH means 1e18 wei
```

>     - **target** - The identifier of the receiver for this transfer. This should be the identifier of the recipient address as specified in `addresses`.

Reviewer: ```suggestion
    - **target** - A string identifying the account to receive tokens, or to adjust allowance for. The string corresponds to a key in the `addresses` section.
```

> **Addresses** - This specifies the addresses of the Gnosis Safe, tokens, and recipients relevant to this template.

Reviewer: ```suggestion
**Addresses** - This specifies the addresses of the Gnosis Safe, tokens, and recipients relevant to this task.
```

 
 - Addresses Object structure:
 
    - **Key** - The identifier of the address. For the Safe that is being used to send tokens, this key is `SafeToSendFrom`. The recipient and token addresses should be assigned names that accurately identify each recipient or token.
    - **Value** - The address corresponding to the identifier's address. For the `SafeToSendFrom` key, this should be the address of the Gnosis Safe that is sending the tokens.

The script can then be run using the following command with the path to the `.toml` file, the network to run the script on, and the block number to fork from filled in:

```bash
forge script src/improvements/template/FinanceTemplate.sol --sig "simulateRun(string)" <path-to-finance-template.toml> --rpc-url <task-network> --fork-block-number <pinned-block-number> -vv
```

This script can be run on testnet using the following command:

```bash
forge script src/improvements/template/FinanceTemplate.sol --sig "simulateRun(string)" test/tasks/mock/configs/TestFinanceTemplate.toml --rpc-url sepolia --fork-block-number 7880546 -vvv

> ```

Reviewer: Is this right? I think we actually need to link to `SINGLE.md` or `NESTED.md` here and use the standard flow
